### DISCUSSION TITLE
AdjustQueryParameters

### DISCUSSION THREAD

**Discussion by Flora-Feng0532** [rdkit/rdkit#4336] [Q&A] [answered]
I got into a problem when trying to do substructure matching of scaffolds:

Scaffold 1: O=c1oc2ccccc2[nH]1
Scaffold 2: c1(ocn2)c2cccc1
Molecule: O=S(N1CCN(C1)c2ncncc2)(c(cc3)cc4c3oc(=O)[nH]4)=O

What I want:
Molecule.HasSubstructureMatch(scaffold 1) --> True
Molecule.HasSubstructureMatch(scaffold 2) --> False

In fact, Molecule believes it has both patterns.

I tried many ways, like ignore explicit Hs, set adjustDegree=True in Chem.AdjustQueryParameters()...
It didnt work unless I was able to define H0(hydrogen zero) in SMARTS strings.
I was hopping is there any solution to this situation?

**Comment by ptosco** [comment]
@Flora-Feng0532 the problem is that `scaffold2` is a different tautomer.
```
from rdkit import Chem
from rdkit.Chem.Draw import MolsToGridImage
from rdkit.Chem.MolStandardize import rdMolStandardize

mol = Chem.MolFromSmiles("O=S(N1CCN(C1)c2ncncc2)(c(cc3)cc4c3oc(=O)[nH]4)=O")
mol
```
![image](https://user-images.githubusercontent.com/5244385/125460037-16265f13-cf75-475f-b7a1-9c5e06a1ba66.png)
```
scaffold1 = Chem.MolFromSmiles("O=c1oc2ccccc2[nH]1")
scaffold2 = Chem.MolFromSmiles("c1(ocn2)c2cccc1")
scaffold1
```
![image](https://user-images.githubusercontent.com/5244385/125460196-c16b2fc3-ac17-4464-9f88-58bf0c6ebaf2.png)
```
scaffold2
```
![image](https://user-images.githubusercontent.com/5244385/125460306-5782cf74-73a9-4248-a809-64881a7578cf.png)

`mol` can actually exist as two tautomers:
```
te = rdMolStandardize.TautomerEnumerator()
tautomers = te.Enumerate(mol)
MolsToGridImage(tautomers)
```
![image](https://user-images.githubusercontent.com/5244385/125460738-c74bc8dd-9544-4176-9bd6-c71756c34460.png)
You can load both tautomers into a `MolBundle` and use that instead of your `mol` for the substructure search, so it will account for both tautomers:
```
tautomer_bundle = Chem.MolBundle()
for t in tautomers:
    tautomer_bundle.AddMol(t)
tautomer_bundle.HasSubstructMatch(scaffold1)
True
tautomer_bundle.HasSubstructMatch(scaffold2)
True
```
As expected, the `MolBundle` will match both scaffold. Effectively, it is attempting a substructure matches with all molecules in the bundle and return `True` as soon as any of the molecules in the bundle matches,

**Reply by Flora-Feng0532** [reply]
:)  It looks like MolBundle is actually useful in tautomers!
But I expect False in scaffold 2 & True in scaffold 1. 
So is there any solution to this ?
Looking forward to your reply

**Comment by ptosco** [comment, marked as answer]
@Flora-Feng0532 Oops, I read your question too quickly and I gave an answer to the opposite problem.
For your problem you might add a total degree query on aromatic heteroatoms if you wish to match a specific tautomer, e.g.:

```
def add_total_degree_query(scaffold):
    with Chem.RWMol(scaffold) as scaffold_q:
        for a in scaffold_q.GetAtoms():
            if (a.GetAtomicNum() != 6 and a.GetIsAromatic()):
                symbol = a.GetSymbol().lower()
                d = a.GetTotalDegree()
                smarts_atom = Chem.MolFromSmarts(f"[{symbol}X{d}]").GetAtomWithIdx(0)
                scaffold_q.ReplaceAtom(a.GetIdx(), smarts_atom)
    return scaffold_q

scaffold1_q = add_total_degree_query(scaffold1)
scaffold2_q = add_total_degree_query(scaffold2)
mol.HasSubstructMatch(scaffold1_q)
True
mol.HasSubstructMatch(scaffold2_q)
False
```

**Reply by Flora-Feng0532** [reply]
Amazing!  Thx a lot!
